packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp

```diff
@@ -115,12 +119,81 @@ namespace MueLu {
       const RCP<const Vector> D = (!fixing ? Utilities::GetInverse(diag) : Utilities::GetInverse(diag, 1e-4, one));
       Ainv = MatrixFactory::Build(D);
     }
+    else if(method=="spai") {
+      Ainv = GetSparseInverse(A, A->getCrsGraph());
+    }
 
     GetOStream(Statistics1) << "Approximate inverse calculated by: " << method << "." << std::endl;
+    GetOStream(Statistics1) << "Ainv has " << Ainv->getGlobalNumRows() << "x" << Ainv->getGlobalNumCols() << " rows and columns." << std::endl;
 
     Set(currentLevel, "Ainv", Ainv);
   }
 
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>>
+  InverseApproximationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetSparseInverse(const RCP<Matrix>& Aorg, const RCP<const CrsGraph>& sparsityPattern) const {
+
+    // construct the inverse matrix with the given sparsity pattern
+    RCP<Matrix> Ainv = MatrixFactory::Build(sparsityPattern);
+    Ainv->resumeFill();
+
+    // gather missing rows from other procs to generate an overlapping map
+    RCP<Import> rowImport = ImportFactory::Build(sparsityPattern->getRowMap(), sparsityPattern->getColMap());
+    RCP<Matrix> A = MatrixFactory::Build(Aorg, *rowImport);
+
+    // loop over all rows of the inverse sparsity pattern (this can be done in parallel)
+    for(size_t k=0; k<sparsityPattern->getLocalNumRows(); k++) {
+
+      // 1. get column indices Ik of local row k
+      ArrayView<const LO> Ik;
+      sparsityPattern->getLocalRowView(k, Ik);
+
+      // 2. get all local A(Ik,:) rows
+      Array<ArrayView<const LO>> J(Ik.size());
+      Array<ArrayView<const SC>> Ak(Ik.size());
+      Array<LO> Jk;
+      for (size_t i = 0; i < Ik.size(); i++) {
+        A->getLocalRowView(Ik[i], J[i], Ak[i]);
+        for (size_t j = 0; j < J[i].size(); j++)
+          Jk.append(J[i][j]);
+      }
+      // set of unique column indices Jk
+      std::sort(Jk.begin(), Jk.end());
+      Jk.erase(std::unique(Jk.begin(), Jk.end()), Jk.end());
+      // create map
+      std::map<LO, LO> G;
+      for (size_t i = 0; i < Jk.size(); i++) G.insert(std::pair<GO, GO>(Jk[i], i));
+
+      // 3. merge rows together
+      Teuchos::SerialDenseMatrix<LO, SC> localA(Jk.size(), Ik.size(), true);
+      for (size_t i = 0; i < Ik.size(); i++) {
+        for (size_t j = 0; j < J[i].size(); j++) {
+          localA(G.at(J[i][j]), i) = Ak[i][j];
+        }
+      }
+
+      // 4. get direction-vector
+      // diagonal needs an entry!
+      Teuchos::SerialDenseVector<LO, SC> ek(Jk.size(), true);
+      ek[std::find(Jk.begin(), Jk.end(), k) - Jk.begin()] = 1.0;
+
+      // 5. solve linear system for x
+      Teuchos::SerialDenseVector<LO, SC> localX(Ik.size());
+      Teuchos::SerialQRDenseSolver<LO, SC> qrSolver;
+      qrSolver.setMatrix(Teuchos::rcp(&localA, false));
+      qrSolver.setVectors(Teuchos::rcp(&localX, false), Teuchos::rcp(&ek, false));
+      qrSolver.solve();
```

Comment:
Is there any error checking to be done or is this guaranteed to always deliver a usable solution?

Reply:
I can include an error check.